### ISSUE TITLE
RFE: new Coordinate(<coordinate>) with metod Coordinate::copy()

### ISSUE THREAD

**Issue by jandam** [locationtech/jts#436]
Please replace all usages of Coordinate copy constructor with method Coordinate::copy.

Reason is to allow preserve all information (instance type and field values) from derived Coordinate type. For example I have following class:
```
public class CoordinateWithID extends Coordinate {
  private int id;
...
}

```
I'm using Noder with modified LineItersector that generates CoordinateWithID and my instance is lost at https://github.com/locationtech/jts/blob/b8c6e88cd4cf220782c326ae6a48dcdfbc38e068/modules/core/src/main/java/org/locationtech/jts/noding/NodedSegmentString.java#L145 due to copy constructor.

**Comment by dr-jts** [comment]
Good idea.  Copy constructor is used in a lot of places, so may take a bit to make this change.

**Comment by dr-jts** [comment]
Actually doesn't this address only part of the problem?  What about net new coordinates which are created?  E.g. when noding is carried out?

This points to a deeper problem in JTS, which the recent coordinate dimension improvements are going to encounter as well.   For instance, if M values are present then ideally they will be preserved, and new Coordinates created with M values (and populated by interpolation).

I'm not sure what a solution to this would involve. Perhaps a pluggable `CoordinateFactory` attached to `GeometryFactory`?  Although I'm not happy about the extra complexity of this, and it seems to have quite a bit of overlap with `CoordinateSequenceFactory`.

**Comment by jandam** [comment]
Yes you are right, it only solves part of problem. It doesn't solve creating and interpolating new coordinates. So using copy method can be considered as code cleanup.
  Otherwise I think that introducing CoordinateFactory is not necessary. CoordinateSequenceFactory can be extended with default method createCoordinate.

```
public interface CoordinateSequenceFactory {
  ...
  default Coordinate createCoordinate() {
    return new Coordinate();
  }
}
```
  My idea about this rfe was to keep information about source geometry during processing in Noder. But I found more elegant solution - NodedSegmentString has field for user data so I copy user data from input geometry and at the end I copy this user data to final line strings.

**Comment by dr-jts** [comment]
> CoordinateSequenceFactory can be extended with default method createCoordinate.

That is simplest, for sure.  One problem is that CoordinateSequenceFactory does not know what dimension of Coordinate is required for any given CoordinateSequence it has created.  So it would have to create the largest Coordinate possible. So in your example, a CoordinateXYZM with Id, and in JTS in the future probably a CoordinateXYZM (which is my current best thinking about the M problem anyway).

**Comment by dr-jts** [comment]
> My idea about this rfe was to keep information about source geometry during processing in Noder. But I found more elegant solution - NodedSegmentString has field for user data so I copy user data from input geometry and at the end I copy this user data to final line strings.

Good - exactly what that capability was intended for.
